lib/IRGen/IRGenSIL.cpp

```diff
@@ -2074,31 +2098,69 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
       assert(0 && "TensorAttributes cannot be dynamic");
     case GraphOperationInfo::ArgumentLowering::Input: {
       assert(argumentName.empty() && "inputs cannot have names");
+
+      // Special attr to represent the list length, if needed. This is necessary
+      // because we add list inputs one element at a time rather than by using a
+      // TFE_OpAddInputList function (which doesn't exist yet). This is null
+      // when the input does not have a number attr.
+      const char *numberAttr = inputListNumberAttr(opInfo.getOperationName(),
+                                                   inputIdx);
+
+      // Keep track of the list length when we have a number attr.
+      llvm::Value *listLength = nullptr;
+      auto incrementListLength = [&](llvm::Value *amount) {
```

Comment:
the code here feels pretty tricky...

one option to simplify it is: make incrementListLength only callable from a list context, and make the caller initialize the list length first before calling this closure, so that here we only do `Builder.CreateAdd(listLength, amount);`

also, i wonder if the logic is correct if we have mixed input params with handles and lists. say the tfop takes 2 params: (list, handle). IIUC, say we have the N attr, and user writes: `#tfop("SomeOp", x, [y, z])`, where x, y and z are tensor handles (this example does not involve the new TensorGroup). In this case, will `listLength` end up being 3, instead of 2?

Reply:
I simplified it to construct the value immediately, so that the closure only has to CreateAdd.

I cannot restrict it to only the SAK_List case, because once I add support for runtime arrays, the runtime arrays will show up as SAK_Singles.

The logic should be correct for mixed handles and lists. For example `#tfop("SomeOp", x, [y, z])` gets 2 `structuredArgument`s: one for `x` and one for `[y, z]`. Then it calculates separate `listLength`s for them in separate loop iterations. So it should get `listLength == 1` for `x` and `listLength == 2` for `[y, z]`.

Reply:
SG overall.

> I cannot restrict it to only the SAK_List case, because once I add support for runtime arrays, the runtime arrays will show up as SAK_Singles.

hmm i feel this case might cause confusion. We'll need to take care (e.g. add more comments on the GraphOperationInfo::SAK_Single enum case if needed)